cli/commands/scaffold/action.go

```diff
@@ -0,0 +1,351 @@
+
+	// run boilerplate
+
+	// prepare boilerplate dir
+	boilerplateDir := util.JoinPath(tempDir, util.DefaultBoilerplateDir)
+	// use template dir as boilerplate dir
+	if templateDir != "" {
+		boilerplateDir = templateDir
+	}
+
+	// if boilerplate dir is not found, create one with default template
+	if !files.IsExistingDir(boilerplateDir) {
+		// no default boilerplate dir, create one
+		boilerplateDir, err = os.MkdirTemp("", "boilerplate")
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		dirsToClean = append(dirsToClean, boilerplateDir)
+		if err := os.WriteFile(util.JoinPath(boilerplateDir, "terragrunt.hcl"), []byte(defaultTerragruntTemplate), 0644); err != nil {
+			return errors.WithStackTrace(err)
+		}
+		if err := os.WriteFile(util.JoinPath(boilerplateDir, "boilerplate.yml"), []byte(defaultBoilerplateConfig), 0644); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	// add additional variables
+	vars["requiredVariables"] = requiredVariables
+	vars["optionalVariables"] = optionalVariables
+
+	vars["sourceUrl"] = moduleUrl
+
+	opts.Logger.Infof("Running boilerplate generation to %s", opts.WorkingDir)
+	boilerplateOpts := &boilerplate_options.BoilerplateOptions{
+		TemplateFolder:  boilerplateDir,
+		OutputFolder:    opts.WorkingDir,
+		OnMissingKey:    boilerplate_options.DefaultMissingKeyAction,
+		OnMissingConfig: boilerplate_options.DefaultMissingConfigAction,
+		Vars:            vars,
+
+		NonInteractive: true,
+	}
+	emptyDep := variables.Dependency{}
+	if err := templates.ProcessTemplate(boilerplateOpts, boilerplateOpts, emptyDep); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	opts.Logger.Infof("Running fmt on generated code %s", opts.WorkingDir)
+	if err := hclfmt.Run(opts); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	opts.Logger.Info("Scaffolding completed")
+
+	return nil
+}
+
+// rewriteModuleUrl rewrites module url to git ssh if required
+func rewriteModuleUrl(opts *options.TerragruntOptions, vars map[string]interface{}, moduleUrl string) (*url.URL, error) {
+	var updatedModuleUrl = moduleUrl
+	sourceUrlType := sourceUrlTypeHttps
+	if value, found := vars[sourceUrlTypeVar]; found {
+		sourceUrlType = fmt.Sprintf("%s", value)
+	}
+
+	// rewrite module url
+	scheme, host, path := parseUrl(opts, moduleUrl)
+	// try to rewrite module url if is https and is requested to be git
+	if scheme == "https" && sourceUrlType == sourceUrlTypeGit {
+		gitUser := sourceGitSshUser
+		if value, found := vars[sourceGitSshUserVar]; found {
+			gitUser = fmt.Sprintf("%s", value)
+		}
+		path = strings.TrimPrefix(path, "/")
+		updatedModuleUrl = fmt.Sprintf("%s@%s:%s", gitUser, host, path)
+	}
+
+	parsedModuleUrl, err := terraform.ToSourceUrl(updatedModuleUrl, opts.WorkingDir)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	return parsedModuleUrl, nil
+}
+
+// rewriteTemplateUrl rewrites template url with reference to tag
```

Comment:
NIT: an example before & after in comments would help clarify what this method does.